chore: Persist validation status in archiver

spalladino opened the pull request:
Adds the _pending chain validation status_ to the archiver store, so it is persisted across node restarts. The pending chain validation status is whether the last block seen in the pending chain is valid or not, as in it having valid attestations or not.

Also, this PR fixes the condition in which it's updated: if a new block with the same number and still invalid is detected, the validation status is updated. This is tested in https://github.com/AztecProtocol/aztec-packages/pull/16836.

This PR also changes that shape of the validation result, so instead of storing a full L2 block, we only store the identifiers and attestations needed to invalidate it. It also adds serialization and static `random` methods to some adjacent classes.

spalladino commented on yarn-project/stdlib/src/block/validate_block_result.ts:
```diff
@@ -0,0 +1,122 @@
+export type ValidateBlockNegativeResult =
+  | {
+      valid: false;
+      /** Identifiers from the invalid block */
+      block: L2BlockInfo;
+      /** Committee members at the epoch this block was proposed */
+      committee: EthAddress[];
+      /** Epoch in which this block was proposed */
+      epoch: bigint;
+      /** Proposer selection seed for the epoch */
+      seed: bigint;
+      /** List of committee members who signed this block proposal */
+      attestors: EthAddress[];
+      /** Committee attestations for this block as they were posted to L1 */
+      attestations: CommitteeAttestation[];
+      /** Reason for the block being invalid: not enough attestations were posted */
+      reason: 'insufficient-attestations';
+    }
+  | {
+      valid: false;
+      /** Identifiers from the invalid block */
+      block: L2BlockInfo;
+      /** Committee members at the epoch this block was proposed */
+      committee: EthAddress[];
+      /** Epoch in which this block was proposed */
+      epoch: bigint;
+      /** Proposer selection seed for the epoch */
+      seed: bigint;
+      /** List of committee members who signed this block proposal */
+      attestors: EthAddress[];
+      /** Committee attestations for this block as they were posted to L1 */
+      attestations: CommitteeAttestation[];
+      /** Reason for the block being invalid: an invalid attestation was posted */
+      reason: 'invalid-attestation';
+      /** Index in the attestations array of the invalid attestation posted */
+      invalidIndex: number;
+    };
+
+/** Result type for validating a block attestations */
+export type ValidateBlockResult = { valid: true } | ValidateBlockNegativeResult;
```
The original shape for this type was:

```
/** Subtype for invalid block validation results */
export type ValidateBlockNegativeResult =
  | {
      valid: false;
      block: PublishedL2Block;
      committee: EthAddress[];
      epoch: bigint;
      seed: bigint;
      attestations: BlockAttestation[];
      reason: 'insufficient-attestations';
    }
  | {
      valid: false;
      block: PublishedL2Block;
      committee: EthAddress[];
      epoch: bigint;
      seed: bigint;
      reason: 'invalid-attestation';
      attestations: BlockAttestation[];
      invalidIndex: number;
    };

/** Result type for validating a block attestations */
export type ValidateBlockResult = { valid: true } | ValidateBlockNegativeResult;
```

spalladino commented on yarn-project/archiver/src/archiver/archiver.ts:
Dear reviewer, if you skim through most of the files, stop and pay special attention to this one.

spalladino commented on yarn-project/foundation/src/eth-signature/eth_signature.ts:
```diff
@@ -94,9 +94,7 @@ export class Signature {
   toBuffer(): Buffer {
-    const buffer = serializeToBuffer([this.r, this.s, this.v]);
-    this.size = buffer.length;
-    return buffer;
+    return serializeToBuffer([this.r, this.s, this.v]);
   }
```
The only reason for this change is to make equality assertions less of a headache. Note that `size` is still being cached when queried via `getSize`.

spalladino commented on yarn-project/stdlib/src/block/l2_block_info.ts:
```diff
@@ -0,0 +1,63 @@
+
+import { z } from 'zod';
+
+export type L2BlockInfo = {
+  blockHash?: Fr;
+  archive: Fr;
+  lastArchive: Fr;
+  blockNumber: number;
+  slotNumber: number;
+  txCount: number;
+  timestamp: bigint;
+};
```
The original shape for `BlockInfo` was:

```
export type BlockInfo = {
  blockHash?: string;
  archive: string;
  blockNumber: number;
  slotNumber: number;
  txCount: number;
};
```

I added the `timestamp` and `lastArchive`, and moved it to its own file.

PhilWindle reviewed (approved):
Most of this PR was down to the change in shape of the various types.

PhilWindle commented on yarn-project/archiver/src/archiver/archiver.ts:
```diff
@@ -841,7 +841,9 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
       try {
-        const [processDuration] = await elapsed(() => this.store.addBlocks(validBlocks));
+        const updatedValidationResult =
+          rollupStatus.validationResult === initialValidationResult ? undefined : rollupStatus.validationResult;
```
This confused me at first, I thought we would be storing `undefined`.

spalladino committed:
chore: Persist validation status in archiver

Adds the _pending chain validation status_ to the archiver store, so it is persisted across node restarts. The pending chain validation status is whether the last block seen in the pending chain is valid or not, as in it having valid attestations or not.

Also, this PR fixes the condition in which it's updated: if a new block with the same number and still invalid is detected, the validation status is updated. This is tested in https://github.com/AztecProtocol/aztec-packages/pull/16836.

This PR also changes that shape of the validation result, so instead of storing a full L2 block, we only store the identifiers and attestations needed to invalidate it. It also adds serialization and static `random` methods to some adjacent classes.

Co-authored-by: PhilWindle <philip.windle@gmail.com>
Co-authored-by: Santiago Palladino <santiago@aztec-labs.com>
